Implementing dht storage interface.

aldenml opened the pull request:
This commit is a first step towards a complete pluggable dht storage. Currently is just a refactor and requieres a lot of review.

aldenml commented:
Hi @jpetso, I took care of your comments but the copyright one (so far is more copy&paste than actual original work). Thank you for take it a look.

@arvidn, I'm a little worried that this PR is not the best way to introduce such a refactor, I don't mind redo it step by step (PR by PR).

I have one particular question: These fields `dht_immutable_item::ips` and `dht_immutable_item ::last_seen` I can't see where are they used. What are they for?

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+		// this counts the number of IPs we have seen
+		// announcing this item, this is used to determine
+		// popularity if we reach the limit of items to store
+		bloom_filter<128> ips;
```
I don't think this type is public, it would have to be exported too probably. Ideally the interface could be kept smaller though. I'd prefer if bloom filter remains internal

aldenml replied:
I refactored this popularity related fields to a private type and using them internally in the dht_default_storage. This way, there is no need for a future implementation to know about it.

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+		virtual size_t num_torrents() const = 0;
+		virtual size_t num_peers() const = 0;
+
+		virtual torrent_entry const* lookup_peers(sha1_hash const& info_hash) const = 0;
```
returning a pointer here either enforces a specific implementation, or it requires an implementation to be less efficient by creating a temporary torrent_entry object and copy all peers into it (even though we're only interested in a subset of peers). I think the interface would be more flexible the function was expected to fill in a vector (or maybe just a buffer) of a random subset of peers instead. The most efficient interface would probably be to ask it fill in binary, compact IP-port pairs into a buffer. then it could be written directly into the packet

aldenml replied:
Now, I'm returning the entry the items by out-parameter. The DHT items are already bytes but this torren_entry is used in node in a way that making it byte-oriented will make it more to work with. See the code starting at https://github.com/arvidn/libtorrent/blob/master/src/kademlia/node.cpp#L671.

But I will consider any other change you suggest.

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+	struct TORRENT_EXPORT dht_storage_interface
+	{
+		virtual size_t num_torrents() const = 0;
+		virtual size_t num_peers() const = 0;
```
these are meant to be global counters?

aldenml replied:
This counters are there due to `node::data_size` and a few deprecated code. I made all of them deprecated.

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+
+		virtual torrent_entry const* lookup_peers(sha1_hash const& info_hash) const = 0;
+		virtual void add_peer(sha1_hash const& info_hash, tcp::endpoint endp
+			, std::string const& name, bool seed) = 0;
```
do you think it would make sense to call this announce_peer instead? it would be a bit more consistent with the underlying dht message

aldenml replied:
done

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+	{
+		char sig[item_sig_len];
+		boost::uint64_t seq;
+		ed25519_public_key key;
```
I don't think this type is public either (but I could be wrong). I suppose it's just a wrapper around a byte array though.

aldenml replied:
This type is just above, it's a wrapper around a byte array and used this way in an operator. I can change it, based on your review.

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+		int salt_size;
+	};
+
+	struct count_peers
```
I imagine this helper class as well as the comparator below are only used internally. if so, they should live in a .cpp file in the anonymous namespace

aldenml replied:
done

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+		virtual void put_immutable_item(sha1_hash const& target
+			, dht_immutable_item const& to_add, address addr) = 0;
+
+		virtual dht_mutable_item const* get_mutable_item(sha1_hash const& target) const = 0;
```
the same thing goes here. it's probably not a good idea to return a pointer. If the implementation doesn't want to use dht_mutable_item internally, it would have to allocate one _just_ to be able to return it. It would also need some kind of pool allocator in order to reclaim it at some point in the future (it's unclear exactly how long the returned object has to be valid for). I think a better approach would be to have the function fill something in. Ideally the thing being filled in is close to or exactly what is going out on the wire to avoid unnecessary data conversion and copying.
are there more than one cases to call this function? I imagine it would be convenient if all the crypto and sequence number checks could be done outside of the storage, to separate the responsibility (and make the storage not responsible at all for any security features basically)

aldenml replied:
Along the lines of the `lookup_peers` review comment. This types (and returned data) is used in a high level form inside node.cpp. I think returning them via ref-out parameter avoid the problem of a long living object in heap.

aldenml commented:
@arvidn Thank you for your comments, I will work on it. About my last question: Can I remove `dht_immutable_item::ips` and `dht_immutable_item ::last_seen` if they are not used?

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+		virtual void tick() = 0;
+	};
+
+	class TORRENT_EXPORT dht_default_storage : public dht_storage_interface, boost::noncopyable
```
would this class need to be exported?

aldenml replied:
Removed the `TORRENT_EXPORT` but I need your thoughts on how to hide this type and the few ones I added.

arvidn commented on include/libtorrent/kademlia/node.hpp:
```diff
@@ -341,6 +263,8 @@ typedef std::map<node_id, dht_mutable_item> dht_mutable_table_t;
 	udp_socket_interface* m_sock;
 	counters& m_counters;
+
+	dht_default_storage m_storage;
```
so the storage isn't customizable (yet)?

aldenml replied:
Not yet :)

aldenml replied:
I want to finish this refactor first and then proceed to the parametrization in the same style of the torrent storage constructor function.

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,223 @@
+		dht_default_storage(sha1_hash const& id, libtorrent::dht_settings const& settings
+			, counters& cnt);
+
+		~dht_default_storage() {}
```
even though the destructor looks empty, there's a lot of things happening in here that probably shouldn't be inlined. (i.e. even though it's empty, the implementation should be put in the .cpp file)

aldenml replied:
done

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1250,17 +1087,13 @@ void node::incoming_request(msg const& m, entry& e)
 			{
 				// this is the case where we already 
-				dht_mutable_item* item = &i->second;
+				dht_mutable_item* item = (dht_mutable_item*)i;
```
why the cast? I would expect this to cause a compile error on travis (because of -Werror)

aldenml replied:
done (actually this code is no more).

arvidn commented:
the bloom filter(s) are meant to be used to keep track of how many _unique_ announcers there are for an item. This is used to rank importance when deciding what item to evict if the table is full. It's possible that it got broken at some point.

aldenml commented:
If you point me to some place where I can look at it, I could restore the feature. It's used in the touch item (https://github.com/arvidn/libtorrent/pull/123/files#diff-99a5e281466aead1cb181c935aa4d7b0R64) to write it though.

arvidn commented:
it's used here: https://github.com/arvidn/libtorrent/blob/master/src/kademlia/node.cpp#L1309
it's used to maintain the num_announcers counter correctly (i.e. not counting duplicates)

aldenml commented:
Thank you, I missed that, indirectly it's used in the `immutable_item_comparator`. I will address your comments.

arvidn commented:
in general I'm concerned about the size of the API. I'd like to keep this addition to a minimum of additional exported types and functions (within reason)

aldenml commented:
I want to point out that the logic  of keys validation during the "pub mutable item" is still done in node.cpp. A dht_storage_interface implementation is (and should be only) for storage, with the additional tick in case some invalidation (and removal) is in order.

@arvidn Waiting for your next review and comments.

arvidn commented:
I'm aiming for another review today. would you mind rebasing on top of master as well? (all builds should be passing now)

aldenml commented:
done

arvidn commented:
Did you see the link error?

https://ci.appveyor.com/project/arvidn/libtorrent/build/1.0.600/job/6ks2nch3x24tkbj8#L3486

arvidn commented:
and the test failures:

https://travis-ci.org/arvidn/libtorrent/jobs/79136578#L919

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,216 @@
+		virtual void tick() = 0;
+	};
+
+	// NON-PUBLIC TYPES
```
ideally they can be moved to the .cpp file (in an anonymous namespace), or at least some of them.

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,216 @@
+		dht_popularity_item popularity;
+	};
+
+	class dht_default_storage : public dht_storage_interface, boost::noncopyable
```
once the storage in the node class is a dht_storage_interface, the concrete type of the default storage can be hidden away inside dht_storage.cpp (or possibly some other .cpp file)

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,355 @@
+	}
+
+	// internal
+	inline bool operator<(ed25519_public_key const& lhs, ed25519_public_key const& rhs)
```
I don't think these have to be inline anymore. also, they should probably be in the anonymous namespace

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,355 @@
+		touch_item(&i->second.popularity, addr);
+	}
+
+	bool dht_default_storage::get_mutable_item(sha1_hash const& target
```
this may be a bit premature for v1 of this patch, but do you have any ideas of how the read operations could be made asynchronous? I'm worried about disk I/O blocking the main thread of libtorrent when using sqlite3 as the backend.

aldenml replied:
No ideas at this moment.

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,355 @@
+			purge_peers(t.peers);
+
+			// if there are no more peers, remove the entry altogether
+			if (t.peers.empty())
```
personally I prefer the style of: if (!t.peers.empty()) continue;

that way you save an indentation step below

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1345,24 +1165,20 @@ void node::incoming_request(msg const& m, entry& e)
 
-		dht_immutable_table_t::iterator imutable_it = m_immutable_table.end();
-
+		dht_immutable_item immutable_it;
```
this is no longer an iterator, but it's still called _it.

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1345,24 +1165,20 @@ void node::incoming_request(msg const& m, entry& e)
-			imutable_it = m_immutable_table.find(target);
-
-		if (imutable_it != m_immutable_table.end())
+		if (!msg_keys[0] && m_storage.get_immutable_item(target, immutable_it))
```
would it make sense to make this fill in an `entry` object which then can be assigned to reply["v"]? That way dht_immutable_item can be private and not part of the interface

aldenml replied:
I need to say that this seems to me like a very good idea. It will take me a little while to rework the API.

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1345,24 +1165,20 @@ void node::incoming_request(msg const& m, entry& e)
 		{
-			dht_mutable_table_t::iterator mutable_it = m_mutable_table.find(target);
-			if (mutable_it != m_mutable_table.end())
+			dht_mutable_item mutable_it;
```
this could just be called `f` and not have a reference create to point at it

arvidn commented on Jamfile:
```diff
@@ -701,6 +701,7 @@ SOURCES =
 	;
 
 KADEMLIA_SOURCES =
+	dht_storage
```
could you add this file to src/Makefile.am too?

aldenml commented:
Thanks @arvidn, I will address your comments and the new rules for PRs.

aldenml commented:
Hi @arvidn, this checks are taking forever. Maybe you want to take a look at the interface `dht_storage_interface` to see if you think it's a better API now. Thanks.

arvidn commented:
maybe I should cut the number of build configurations on appveyor, since it's especially slow.

aldenml commented:
Have you tried the trial of appveyor?

I managed to get the `dht_storage_interface` as plain as possible. No extra types, only primitives and libtorrent::entry. I will wait for your review (and fix any check that does not get green).

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -115,9 +98,16 @@ node::node(udp_socket_interface* sock
+}
+
+node::~node() {
+	if (m_storage != NULL) delete m_storage;
```
I would much rather see m_storage be a boost::scoped_ptr and not manually destruct it in the destructor. (btw, delete NULL; is fine, there's no need to check for it).

arvidn replied:
in this case I don't believe node is copyable/movable or assignable. but in the general case, having raw pointers as members would likely cause problems with those operations and are code smells

aldenml replied:
done.

Thanks for the comment, I learned about "scoped_ptr" and learning more about smart pointers.
I remember your other comment about the "delete NULL", I suppose is a bad habit due to some bad experience with a badly rewritten malloc/free|new/delete.

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,537 @@
+/*
+
+Copyright (c) 2012-2015, Arvid Norberg
```
I believe you should add your name here too

aldenml replied:
done

arvidn commented:
looks good. and the fact that it passes builds and tests under all the CI configs makes me fairly confident that it's correct as well. see comment though, I'd rather not have the raw pointer member

aldenml commented:
Good. Green.

arvidn commented:
there's a challenge of documentation as well. currently no type in the dht is documented (documentation is not generated from headers under kademlia iirc).

arvidn commented:
in the main libtorrent headers there's a distinction (that I'm working on making clearer) between internal and public declarations and definitions (anything in libtorrent::aux is internal). There isn't quite the same distinction in the dht namespace. If you're using the TORRENT_EXPORT macro, that may be enough though.

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,532 @@
+		}
+	}
+
+	dht_storage_interface* dht_default_storage_constructor(sha1_hash const& id
```
I suppose boost::scoped_ptr can't be copied, so we have to resort to a raw pointer here to stay c++98 compatible.

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -668,25 +627,23 @@ void node::lookup_peers(sha1_hash const& info_hash, entry& reply
-	if (i->first != info_hash) return;
+	std::string v_name;
+	std::vector<std::pair<tcp::endpoint, bool> > v_peers;
+	if (!m_storage->lookup_peers(info_hash, v_name, v_peers)) return;
```
I just realized this: If lookup peers returns a subset of peers (say, no more than 50), then the scrape result below is incorrect, since it assumes v_peers is the complete list. It's a bit unfortunate. the two options I can see

aldenml replied:
I was expecting this comment (and the one below).

Thinking about it, I think one solution is making the lookup_peers similar to the others. You pass the entry, and it's filled with the specified fields. I will change it and you tell me.

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -695,18 +652,18 @@ void node::lookup_peers(sha1_hash const& info_hash, entry& reply
-		int num = (std::min)(int(v.peers.size()), m_settings.max_peers_reply);
-		std::set<peer_entry>::const_iterator iter = v.peers.begin();
+		int num = (std::min)(int(v_peers.size()), m_settings.max_peers_reply);
+		std::vector<std::pair<tcp::endpoint, bool> >::const_iterator iter = v_peers.begin();
```
maybe you can add a comment somewhere indicating that the bool here indicates whether the peer is a seed or not.

arvidn replied:
actually, what would you think about passing in "noseed" to dht_storage::lookup_peers()? That way we don't have to return _all_ peers there (which could get expensive)

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1210,111 +1062,58 @@ void node::incoming_request(msg const& m, entry& e)
 			VALGRIND_CHECK_MEM_IS_DEFINED(msg_keys[4].string_ptr(), item_sig_len);
 			VALGRIND_CHECK_MEM_IS_DEFINED(pk, item_pk_len);
 #endif
+			boost::uint64_t seq = boost::uint64_t(msg_keys[2].int_value());
```
I think this is a dangerous cast. If someone announces a negative sequence number, you will interpret it as positive here (and presumably more recent than the authentic item). I would think it's safer to consistently use signed int64_t for sequence numbers

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1210,111 +1062,58 @@ void node::incoming_request(msg const& m, entry& e)
 
-//				fprintf(stderr, "added mutable item (%d)\n", int(m_mutable_table.size()));
+			boost::uint64_t item_seq;
+			if (!m_storage->get_mutable_item_seq(target, item_seq))
```
presumably this would have to take int64_t also

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1210,111 +1062,58 @@ void node::incoming_request(msg const& m, entry& e)
 					return;
 				}
 
-				if (item->seq > boost::uint64_t(msg_keys[2].int_value()))
```
and clearly I'm guilty of this mistake here

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1349,32 +1148,21 @@ void node::incoming_request(msg const& m, entry& e)
-			{
-				dht_mutable_item const& f = mutable_it->second;
-				reply["seq"] = f.seq;
-				if (!msg_keys[0] || boost::uint64_t(msg_keys[0].int_value()) < f.seq)
```
and here

arvidn commented:
it's starting to take good shape

aldenml commented:
Hi @arvidn, I changed the interface (and implementation) to work with and entry in lookup_peers and signed integer for the mutable sequence number (also added the check with default code error number).

About the documentation. Do you think documenting `dht_storage_interface` is enough?

I'm still not sure about `dht_default_storage_constructor` and the counters parameters, I'm delaying that to when the thing become real in the session API.

arvidn commented:
I would imagine documenting the interface would be enough. I don't know if you've had a chance to look at how the documentation is generated (it's not pretty! but I just can't stand doxygen/javadoc comments). I imagine the kademlia header directory may have to be added to the python script to pick up the dht_storage_interface class documentation.

arvidn commented:
we can fix the documentation separately though. Also, how would you feel about squashing some of your commits? (especially things like fixing typos and juggling trying to get builds to pass).

aldenml commented:
Squash it is.

aldenml commented:
Squashed and documented the interface. Travis is green but appveyor is queued forever.

arvidn commented:
the appveyor failures are because the builds are taking too long (or, appveyor being very slow and having a short timeout). I've fixed master to build quicker and one less configuration. If you rebase now I believe it will pass

aldenml commented:
I see this error: ..\src\kademlia\dht_storage.cpp(310): error C3861: 'random': identifier not found

I will include random.hpp just in case.

aldenml commented:
Green again (appveyor should be used for some kind of "proof of patience").

arvidn commented:
I'm going through this one last time and I'm going to annotate some minor nit-picks. I can fix them after merging the patch though.

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+	//
+	// The default storage implementation uses three maps in RAM to save
+	// the peers, mutable and immutable items and it's designed to
+	// provide a fast and fully complaint behavior of the BEPs.
```
s/complaint/compliant

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+	// provide a fast and fully complaint behavior of the BEPs.
+	//
+	// libtorrent comes with one built-in storage implementation:
+	// ``dht_default_storage`` (private non-accessible class). Their
```
s/Their/its

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+		// peers["BFsd"] - with the standard bit representation of a
+		//                 256 bloom filter containing the seeders
+		//
+		// If no scraping, you should fill the key peers["values"] with
```
the ``scrape`` parameter should probably be referred to by name here

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+		virtual bool lookup_peers(sha1_hash const& info_hash
+			, bool noseed, bool scrape
+			, entry& peers) const = 0;
+		// This method is named announce_peer for consistency with the
```
the c++ terminology is "member function"

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+		// returns true if an entry with the info_hash is found and
+		// the data is returned inside the (entry) out parameter peers.
+		//
+		virtual bool lookup_peers(sha1_hash const& info_hash
```
I wonder if this function should be called get_peers() (for symmetry with the protocol)

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+		// responsibility is store the peer in such a way that it's returned
+		// in the entry with the lookup_peers.
+		//
+		// The name parameter is the name of the torrent if provided during
```
s/name/``name``

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+			, tcp::endpoint const& endp
+			, std::string const& name, bool seed) = 0;
+
+		// This function retrieves the immutable item given the target.
```
s/the target/its target hash

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+		virtual bool get_immutable_item(sha1_hash const& target
+			, entry& item) const = 0;
+		// Store the item's data. This layer is only for storage, the
+		// checking of the target info_hash and the data is performed in
```
"info_hash" isn't correct here. The correct description may be a bit involved (since it's the hash of the public key + sequence number + salt and involves the signature). Perhaps it could just say:

"The authentication of the item is performed by the upper layer."

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+			, boost::int64_t seq, bool force_fill
+			, entry& item) const = 0;
+		// Store the item's data. This layer is only for storage, the
+		// checking of the target info_hash and the valid signature is
```
same thing here about info_hash

arvidn commented on include/libtorrent/kademlia/dht_storage.hpp:
```diff
@@ -0,0 +1,200 @@
+		// performed in upper layers.
+		//
+		// For implementers:
+		// The sequence number should be checked if the item is already
```
it's unclear what "checked" means here. it looks a bit like a layer violation if a bug in the storage could result in a security issue. (i.e. failure to compare the sequence number)

aldenml replied:
Yes, "checked" was a placeholder for some explanation I don't now how clearly express. I appreciate if you put something here.

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,568 @@
+		};
+	}
+
+	class dht_default_storage : public dht_storage_interface, boost::noncopyable
```
this class could also be put in the anonymous namespace I would imagine.

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,568 @@
+		dht_mutable_table_t m_mutable_table;
+	};
+
+	dht_default_storage::dht_default_storage(sha1_hash const& id
```
for classes that are private in a translation unit, I normally put all the implementations inline (to not have to write the signatures twice).

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,568 @@
+		~dht_default_storage();
+
+#ifndef TORRENT_NO_DEPRECATE
+		size_t num_torrents() const;
```
all functions implementing the dht_storage_interface should be marked with TORRENT_OVERRIDE

arvidn commented on src/kademlia/dht_storage.cpp:
```diff
@@ -0,0 +1,568 @@
+
+		dht_mutable_item const& f = i->second;
+		item["seq"] = f.seq;
+		if (force_fill || (0 <= seq && seq < f.seq))
```
this expected behavior should probably be described in more detail in the documentation (in dht_storage_interface)

aldenml replied:
I think it's already documented, but maybe in a cryptic form.

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1210,111 +1019,65 @@ void node::incoming_request(msg const& m, entry& e)
 #endif
+			boost::int64_t seq = msg_keys[2].int_value();
+
+			if (seq < 0)
```
nice

arvidn commented on src/kademlia/node.cpp:
```diff
@@ -1210,111 +1019,65 @@ void node::incoming_request(msg const& m, entry& e)
 
-//				fprintf(stderr, "added mutable item (%d)\n", int(m_mutable_table.size()));
+			boost::int64_t item_seq;
+			if (!m_storage->get_mutable_item_seq(target, item_seq))
```
this will cause two lookups right now, unless the storage implementation has a cache of the last lookup. I can't think of a better way though

aldenml replied:
I suppose this is one of those cases of abstractions taking a toll in performance. However, and only to comment about it, the double lookup is only evident if we know how the current default storage is implemented. The semantic of `get_mutable_item_seq` and `put_mutable_item` taken together does not imply the need of the double lookup.

aldenml commented:
I took care of your comments and suggestions.

aldenml commented:
I need some help with the first task in appveyor

arvidn commented:
hm. this appears to have coincided with appveyor installing boost-1.59. it seems like the msvc-14 installation got (more) broken.

arvidn commented:
merge failed. could you rebase?

aldenml commented:
Rebased, but still CI appveyor test are queued and one travis is failing.

aldenml commented:
Just to document, the CI test are failing randomly.

arvidn commented:
getting: "Head branch was modified. Review and try the merge again."

aldenml commented:
Let me review that.

aldenml commented:
Green (and tested merge in my master branch). Let me know.
